File: site/src/pipeline/signal.ts
// Vehicle signal phase from the three-lamp head on the median nose (signal.py): a box per
// lamp, the colour contrast inside it, the phase per frame and the offline clean-up.
// The Python frame is BGR; here pixels are RGB or RGBA as a canvas or decoder gives them.
// CausalPhase is not ported: analyse() uses fill_phases, and nothing else reads the signal online.

import { percentile, roundHalfEven, warpPoints, type Mat3 } from "./geometry.ts";

export type Phase = "red" | "yellow" | "green" | "off" | "unknown";
export type Box = [number, number, number, number];

export const LAMPS = ["red", "yellow", "green"] as const;
const CODES: Record<Phase, number> = { red: 0, yellow: 1, green: 2, off: 3, unknown: 4 };
const NAMES: Phase[] = ["red", "yellow", "green", "off", "unknown"];

/**
 * Work-pixel boxes [x0, y0, x1, y1) around the red, yellow and green lamp centres. Pass
 * matInv(H) as it comes, like np.linalg.inv(alignment.H): the radius scales with the
 * determinant of its top-left 2x2, so renormalising it would change the boxes slightly.
 */
export function lampPatches(HrefToWork: Mat3, lamps: Record<"red" | "yellow" | "green", [number, number]>, radiusRef = 2.5): Box[] {
  const scale = Math.sqrt(Math.abs(HrefToWork[0] * HrefToWork[4] - HrefToWork[1] * HrefToWork[3]));
  const r = Math.max(1.0, radiusRef * scale);
  return LAMPS.map((name) => {
    const [[x, y]] = warpPoints([lamps[name]], HrefToWork);
    return [roundHalfEven(x - r), roundHalfEven(y - r), roundHalfEven(x + r) + 1, roundHalfEven(y + r) + 1];
  });
}

/**
 * Colour contrast of each lamp: [redness of top, warmth of middle, greenness of bottom].
 * `pixels` is a crop of the work frame whose top-left corner is at `origin` (work pixels);
 * boxes are clipped to the crop, so it must hold them whole to match Python on the full frame.
 */
export function lampScores(
  pixels: Uint8Array | Uint8ClampedArray,
  width: number,
  height: number,
  channels: 3 | 4,
  boxes: Box[],
  origin: [number, number] = [0, 0],
): number[] {
  const out = [0, 0, 0];
  boxes.forEach(([bx0, by0, bx1, by1], k) => {
    const x0 = Math.max(0, bx0 - origin[0]), x1 = Math.min(width, bx1 - origin[0]);
    const y0 = Math.max(0, by0 - origin[1]), y1 = Math.min(height, by1 - origin[1]);
    const v: number[] = [];
    for (let y = y0; y < y1; y++) {
      for (let x = x0; x < x1; x++) {
        const i = (y * width + x) * channels;
        const r = pixels[i], g = pixels[i + 1], b = pixels[i + 2];
        v.push(k === 0 ? r - Math.max(g, b) : k === 1 ? r - b : g - Math.max(r, b));
      }
    }
    if (v.length) out[k] = Math.fround(percentile(v, 75));
  });
  return out;
}

/** The lamp furthest above its own threshold, or "off" when none reaches it (float32 like NumPy). */
export function phaseFromScores(scores: ArrayLike<number>, minContrast: ArrayLike<number>): Phase {
  let k = 0;
  let best = -Infinity;
  for (let i = 0; i < 3; i++) {
    const ratio = Math.fround(Math.fround(scores[i]) / Math.fround(minContrast[i]));
    if (ratio > best) [best, k] = [ratio, i];
  }
  return best >= 1.0 ? LAMPS[k] : "off";
}

/**
 * Offline clean-up: 1-sample blips removed, short dark spells take the phase before them.
 * Flashing green shows up as green/off alternation, so it stays green; a dark spell longer
 * than maxOff seconds becomes unknown.
 */
export function fillPhases(raw: string[], times: ArrayLike<number>, maxOff = 4): Phase[] {
  const codes = raw.map((p) => CODES[p as Phase]);
  const n = codes.length;
  const OFF = CODES.off;
  for (let i = 1; i < n - 1; i++) {
    // in place, as in Python: a flip fixed at i is what i + 1 compares with
    if (codes[i - 1] === codes[i + 1] && codes[i + 1] !== codes[i] && codes[i - 1] !== OFF) codes[i] = codes[i - 1];
  }
  const out = [...codes];
  let i = 0;
  while (i < n) {
    if (codes[i] !== OFF) {
      i++;
      continue;
    }
    let j = i;
    while (j < n && codes[j] === OFF) j++;
    const prev = i > 0 ? codes[i - 1] : j < n ? codes[j] : CODES.unknown;
    const span = j > i ? times[j - 1] - times[i] : 0;
    const fill = span <= maxOff ? prev : CODES.unknown;
    for (let k = i; k < j; k++) out[k] = fill;
    i = j;
  }
  return out.map((c) => NAMES[c]);
}

File: src/parivision/signal.py
"""Vehicle signal phase from the three-lamp head on the median nose.

Each lamp gets a small patch around its centre (mapped from reference pixels
into the frame). A lit red lamp is much redder than its dark housing, a lit
green one much greener, and a lit yellow one much warmer. Each lamp's contrast
is divided by its own threshold (``MIN_CONTRAST``) and the lamp furthest above
it wins; when every lamp is below its threshold the head is "off"
(the flashing-green part of the cycle, or a car hiding the head). Lamp
position, not just colour, decides the phase, so tail lights passing behind
the head do not read as red.

Sequence at this junction: green (the last 3 s flashing), 3 s yellow, red,
then back to green. The afternoon plan lights red and yellow together for the
last 3 s of red.
"""
from __future__ import annotations

import cv2
import numpy as np

from .scene import SIGNAL_LAMPS

RED, YELLOW, GREEN, OFF, UNKNOWN = "red", "yellow", "green", "off", "unknown"
LAMPS = ("red", "yellow", "green")
_CODES = {RED: 0, YELLOW: 1, GREEN: 2, OFF: 3, UNKNOWN: 4}
_NAMES = {v: k for k, v in _CODES.items()}
# Colour contrast (0-255 scale) each lamp needs to count as lit. The green LED
# looks cyan in sunlight, so its green-minus-red/blue contrast is much lower.
MIN_CONTRAST = np.array([20.0, 20.0, 6.0], np.float32)


def lamp_patches(H_ref_to_frame: np.ndarray, radius_ref: float = 2.5) -> list[tuple[int, int, int, int]]:
    """Frame-pixel boxes around the red, yellow and green lamp centres."""
    boxes = []
    scale = float(np.sqrt(abs(np.linalg.det(H_ref_to_frame[:2, :2]))))
    r = max(1.0, radius_ref * scale)
    for name in LAMPS:
        x, y = SIGNAL_LAMPS[name]
        p = cv2.perspectiveTransform(np.float64([[[x, y]]]), H_ref_to_frame)[0, 0]
        boxes.append((int(round(p[0] - r)), int(round(p[1] - r)), int(round(p[0] + r)) + 1, int(round(p[1] + r)) + 1))
    return boxes


def lamp_scores(frame: np.ndarray, boxes) -> np.ndarray:
    """Colour contrast of each lamp: [redness of top, warmth of middle, greenness of bottom]."""
    out = np.zeros(3, np.float32)
    h, w = frame.shape[:2]
    for k, (x0, y0, x1, y1) in enumerate(boxes):
        patch = frame[max(0, y0):min(h, y1), max(0, x0):min(w, x1)].reshape(-1, 3).astype(np.float32)
        if len(patch) == 0:
            continue
        b, g, r = patch[:, 0], patch[:, 1], patch[:, 2]
        if k == 0:
            v = r - np.maximum(g, b)
        elif k == 1:
            v = r - b
        else:
            v = g - np.maximum(r, b)
        out[k] = float(np.percentile(v, 75))
    return out


def phase_from_scores(scores: np.ndarray, min_contrast: np.ndarray = MIN_CONTRAST) -> str:
    ratio = np.asarray(scores, np.float32) / min_contrast
    k = int(np.argmax(ratio))
    return LAMPS[k] if ratio[k] >= 1.0 else OFF


def fill_phases(raw: list[str], times: np.ndarray, max_off: float = 4.0) -> list[str]:
    """Offline clean-up: 1-sample blips removed, short dark spells take the phase before them.

    Flashing green shows up as green/off alternation, so it stays green; a dark
    spell longer than ``max_off`` becomes unknown.
    """
    codes = [_CODES[p] for p in raw]
    n = len(codes)
    for i in range(1, n - 1):  # isolated single-sample flips
        if codes[i - 1] == codes[i + 1] != codes[i] and codes[i - 1] != _CODES[OFF]:
            codes[i] = codes[i - 1]
    out = list(codes)
    i = 0
    while i < n:
        if codes[i] != _CODES[OFF]:
            i += 1
            continue
        j = i
        while j < n and codes[j] == _CODES[OFF]:
            j += 1
        prev = codes[i - 1] if i > 0 else (codes[j] if j < n else _CODES[UNKNOWN])
        span = (times[j - 1] - times[i]) if j > i else 0.0
        fill = prev if span <= max_off else _CODES[UNKNOWN]
        for k in range(i, j):
            out[k] = fill
        i = j
    return [_NAMES[c] for c in out]
